pkg/reporting/object_writer.go

```diff
@@ -0,0 +1,76 @@
+//
+// Copyright 2020 Chef Software, Inc.
+// Author: Marc Paradise <marc@chef.io>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package reporting
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	chef "github.com/chef/go-chef"
+	"github.com/pkg/errors"
+)
+
+type ObjectWriter struct {
+	RootDir string
+}
+
+type ObjectWriterInterface interface {
+	WriteRole(*chef.Role) error
+	WriteEnvironment(*chef.Environment) error
+	WriteNode(*chef.Node) error
+}
+
+func (ow *ObjectWriter) WriteRole(role *chef.Role) error {
+	return ow.save("roles", role.Name, role)
+}
+
+func (ow *ObjectWriter) WriteEnvironment(env *chef.Environment) error {
+	return ow.save("environments", env.Name, env)
+}
+
+func (ow *ObjectWriter) WriteNode(node *chef.Node) error {
+	return ow.save("nodes", node.Name, node)
+}
+
+func (ow *ObjectWriter) save(objGroupingName string, objName string, object interface{}) error {
```

Comment:
From the interactive review: we do not currently have full test coverage on this file. We could pass in an `os` object that we mock, but then that object needs to get passed all the way down the calling chain.

@marcparadise What if we wrote an integration test for this instead of a unit test? That would at least test the happy path.

Reply:
I'm not sure we can - it needs a chef server environment right? 

Actually, that might be a new card - configure chef-analyze integration tests to launch an instance of chef-zero.  Then all kinds of things will be possible. 

After accepting that I couldn't get around testing that golang's file writing and checking code is working in a unit test, I added a happy-path test to verify that this function writes JSON content to the correct place, and that we can read it back in and have it be the thing we expect.

Reply:
For this PR you did exactly what I was thinking - happy path test that JSON gets written out correctly.